Applying resolution strategies to all configurations does not work when spotbugsMain task is configured

Hi @KengoTODA ,

I found that when you have a build file like this:

```
buildscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2/"
        }
    }
    dependencies {
        classpath "gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin-v2:0.4.4"
    }
}

plugins {
    id 'java'
}

repositories {
    jcenter()
}
allprojects {
    apply plugin: "jp.skypencil.spotbugs.snom"
    
    spotbugsMain {
        reports {
            html {
                enabled = true
                destination = file("$buildDir/reports/spotbugs/main/spotbugs.html")
                stylesheet = 'fancy-hist.xsl'
            }
        }
    }

    configurations.all {
        exclude group: 'ch.qos.logback', module: 'logback-classic'
        exclude group: 'ch.qos.logback', module: 'logback-core'
        exclude group: 'org.apache.logging.log4j', module: 'log4j-to-slf4j'
        exclude group: 'org.slf4j', module: 'slf4j-log4j12'
    }

}
```

When running `./gradlew spotbugsMain` you would get the following error:

```
Caused by: org.gradle.api.InvalidUserDataException: Cannot change dependencies of configuration ':spotbugs' after it has been resolved.
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.preventIllegalMutation(DefaultConfiguration.java:1147)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.validateMutation(DefaultConfiguration.java:1118)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.exclude(DefaultConfiguration.java:906)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.exclude(DefaultConfiguration.java:140)
        at org.gradle.internal.metaobject.BeanDynamicObject$MetaClassAdapter.invokeMethod(BeanDynamicObject.java:483)
        at org.gradle.internal.metaobject.BeanDynamicObject.tryInvokeMethod(BeanDynamicObject.java:195)
        at org.gradle.internal.metaobject.CompositeDynamicObject.tryInvokeMethod(CompositeDynamicObject.java:98)
        at org.gradle.internal.extensibility.MixInClosurePropertiesAsMethodsDynamicObject.tryInvokeMethod(MixInClosurePropertiesAsMethodsDynamicObject.java:34)
        at org.gradle.internal.metaobject.ConfigureDelegate.invokeMethod(ConfigureDelegate.java:56)
```

while re-ordering the `configurations` code block to be before `spotbugsMain` works, one use case is to have a multi-module project where the root build file applies spotbugs to all and then one of the subprojects do changes to the resolution strategies.

This works with current plugin `apply plugin: "com.github.spotbugs"`

Also it seems that the problem is only with `stylesheet`

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/AndroidFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/AndroidFunctionalTest.groovy
@@ -16,10 +16,11 @@ package com.github.spotbugs.snom
 import org.gradle.internal.impldep.com.google.common.io.Files
 import org.gradle.testkit.runner.BuildResult
 import org.gradle.testkit.runner.GradleRunner
-import org.gradle.testkit.runner.TaskOutcome
 import org.junit.jupiter.api.BeforeEach
 import spock.lang.Ignore
 import spock.lang.Specification
+
+import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 import static org.junit.jupiter.api.Assertions.assertEquals
 
 class AndroidFunctionalTest extends Specification {
@@ -96,6 +97,6 @@ public class Foo {
                 .build()
 
         then: "gradle runs spotbugsMain successfully"
-        assertEquals(result.task(":spotbugsMain").outcome, TaskOutcome.SUCCESS)
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
     }
 }

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/ExtensionFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/ExtensionFunctionalTest.groovy
@@ -19,6 +19,7 @@ import org.junit.jupiter.api.BeforeEach
 import spock.lang.Specification
 
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
+import static org.junit.jupiter.api.Assertions.assertEquals
 import static org.junit.jupiter.api.Assertions.assertTrue
 
 class ExtensionFunctionalTest extends Specification {
@@ -71,7 +72,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-include"))
         assertTrue(result.getOutput().contains(filter.getAbsolutePath()))
     }
@@ -94,7 +95,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-exclude"))
         assertTrue(result.getOutput().contains(filter.getAbsolutePath()))
     }
@@ -113,7 +114,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-visitors, FindSqlInjection,SwitchFallthrough,"))
     }
 
@@ -130,7 +131,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-omitVisitors, FindSqlInjection,SwitchFallthrough,"))
     }
 
@@ -147,7 +148,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-onlyAnalyze, com.foobar.MyClass,com.foobar.mypkg.*,"))
     }
 
@@ -165,7 +166,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-nested:false"))
         assertTrue(result.getOutput().contains("-Duser.language=ja"))
     }
@@ -183,7 +184,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-Xmx256m"))
     }
 
@@ -201,7 +202,7 @@ spotbugs {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("-effort:less"))
         assertTrue(result.getOutput().contains("-high"))
     }

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/MultiProjectFunctionalTest.groovy
@@ -78,7 +78,7 @@ repositories {
                 .build()
 
         then:
-        assertEquals(TaskOutcome.SUCCESS, result.task(":sub:classes").getOutcome())
-        assertEquals(TaskOutcome.SUCCESS, result.task(":sub:spotbugsMain").getOutcome())
+        assertEquals(TaskOutcome.SUCCESS, result.task(":sub:classes").outcome)
+        assertEquals(TaskOutcome.SUCCESS, result.task(":sub:spotbugsMain").outcome)
     }
 }

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/ReportFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/ReportFunctionalTest.groovy
@@ -65,7 +65,7 @@ public class Foo {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File reportsDir = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").toFile()
         assertFalse(reportsDir.isDirectory())
     }
@@ -85,7 +85,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File report = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").resolve("main").resolve("spotbugs.txt").toFile()
         assertTrue(report.isFile())
     }
@@ -106,7 +106,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File report = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").resolve("main").resolve("spotbugs.html").toFile()
         assertTrue(report.isFile())
         assertTrue(result.getOutput().contains("-html,"))
@@ -138,7 +138,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File report = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").resolve("main").resolve("spotbugs.html").toFile()
         assertTrue(report.isFile())
         assertTrue(result.getOutput().contains("-html:"))
@@ -167,7 +167,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File report = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").resolve("main").resolve("spotbugs.html").toFile()
         assertTrue(report.isFile())
         assertTrue(result.getOutput().contains("-html:"))
@@ -188,7 +188,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File report = rootDir.toPath().resolve("build").resolve("reports").resolve("spotbugs").resolve("main").resolve("spotbugs.xml").toFile()
         assertTrue(report.isFile())
     }
@@ -212,7 +212,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         File reportsDir = rootDir.toPath().resolve("build").resolve("spotbugs").toFile();
         assertTrue(reportsDir.isDirectory())
         File report = reportsDir.toPath().resolve("main").resolve("spotbugs.txt").toFile()
@@ -248,7 +248,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("Running SpotBugs by Gradle Worker..."));
     }
 
@@ -265,8 +265,8 @@ com.github.spotbugs.snom.worker=false
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
-        assertTrue(result.getOutput().contains("Running SpotBugs by JavaExec..."));
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
+        assertTrue(result.getOutput().contains("Running SpotBugs by JavaExec..."))
     }
 
     def "can run task by JavaExec by commandline option"() {
@@ -279,8 +279,8 @@ com.github.spotbugs.snom.worker=false
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
-        assertTrue(result.getOutput().contains("Running SpotBugs by JavaExec..."));
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
+        assertTrue(result.getOutput().contains("Running SpotBugs by JavaExec..."))
     }
 
     def "does not resolve spotbugs configuration by setting stylesheet"() {

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/ReportFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/ReportFunctionalTest.groovy
@@ -19,6 +19,7 @@ import spock.lang.Specification
 import org.gradle.testkit.runner.GradleRunner
 
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
+import static org.junit.jupiter.api.Assertions.assertEquals
 import static org.junit.jupiter.api.Assertions.assertFalse
 import static org.junit.jupiter.api.Assertions.assertNotNull
 import static org.junit.jupiter.api.Assertions.assertTrue
@@ -280,5 +281,32 @@ com.github.spotbugs.snom.worker=false
         then:
         result.task(":spotbugsMain").outcome == SUCCESS
         assertTrue(result.getOutput().contains("Running SpotBugs by JavaExec..."));
+    }
+
+    def "does not resolve spotbugs configuration by setting stylesheet"() {
+        given:
+        buildFile << """
+spotbugsMain {
+    reports {
+        html {
+            stylesheet = 'fancy-hist.xsl'
+        }
+    }
+}
+
+configurations.spotbugs {
+    exclude group: 'org.slf4j', module: 'slf4j-log4j12'
+}
+"""
+        when:
+        def result = GradleRunner.create()
+                .withProjectDir(rootDir)
+                .withArguments('spotbugsMain')
+                .withPluginClasspath()
+                .forwardOutput()
+                .build()
+
+        then:
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
     }
 }

--- a/src/functionalTest/groovy/com/github/spotbugs/snom/StandardFunctionalTest.groovy
+++ b/src/functionalTest/groovy/com/github/spotbugs/snom/StandardFunctionalTest.groovy
@@ -67,8 +67,8 @@ public class Foo {
                 .build()
 
         then:
-        assertEquals(TaskOutcome.SUCCESS, result.task(":classes").getOutcome())
-        assertEquals(TaskOutcome.SUCCESS, result.task(":spotbugsMain").getOutcome())
+        assertEquals(TaskOutcome.SUCCESS, result.task(":classes").outcome)
+        assertEquals(TaskOutcome.SUCCESS, result.task(":spotbugsMain").outcome)
     }
 
     def "can use the specified SpotBugs version"() {
@@ -87,7 +87,7 @@ dependencies {
                 .build()
 
         then:
-        assertEquals(TaskOutcome.SUCCESS, result.task(":classes").getOutcome())
+        assertEquals(TaskOutcome.SUCCESS, result.task(":classes").outcome)
         assertTrue(result.output.contains("spotbugs-4.0.0-beta4.jar"))
     }
 
@@ -107,7 +107,7 @@ dependencies {
                 .build()
 
         then:
-        assertEquals(TaskOutcome.NO_SOURCE, result.task(":spotbugsMain").getOutcome())
+        assertEquals(TaskOutcome.NO_SOURCE, result.task(":spotbugsMain").outcome)
     }
 
     def "can use effort and reportLevel"() {
@@ -148,7 +148,7 @@ spotbugsMain {
                 .build()
 
         then:
-        result.task(":spotbugsMain").outcome == SUCCESS
+        assertEquals(SUCCESS, result.task(":spotbugsMain").outcome)
         assertTrue(result.getOutput().contains("SpotBugsMain ignores failures? true"))
     }
 
@@ -174,7 +174,7 @@ spotbugsMain {
                 .build()
 
         then:
-        assertEquals(TaskOutcome.UP_TO_DATE, result.task(":classes").getOutcome())
-        assertEquals(TaskOutcome.UP_TO_DATE, result.task(":spotbugsMain").getOutcome())
+        assertEquals(TaskOutcome.UP_TO_DATE, result.task(":classes").outcome)
+        assertEquals(TaskOutcome.UP_TO_DATE, result.task(":spotbugsMain").outcome)
     }
 }

--- a/src/main/groovy/com/github/spotbugs/snom/internal/SpotBugsHtmlReport.java
+++ b/src/main/groovy/com/github/spotbugs/snom/internal/SpotBugsHtmlReport.java
@@ -26,12 +26,14 @@ import org.gradle.api.resources.TextResource;
 
 public class SpotBugsHtmlReport extends SpotBugsReport {
   private final Property<TextResource> stylesheet;
+  private final Property<String> stylesheetPath;
 
   public SpotBugsHtmlReport(ObjectFactory objects, SpotBugsTask task) {
     super(objects, task);
     // the default reportsDir is "$buildDir/reports/spotbugs/${taskName}/spotbugs.html"
     setDestination(task.getReportsDir().map(dir -> new File(dir, "spotbugs.html")));
     stylesheet = objects.property(TextResource.class);
+    stylesheetPath = objects.property(String.class);
   }
 
   @NonNull
@@ -42,7 +44,7 @@ public class SpotBugsHtmlReport extends SpotBugsReport {
     if (stylesheet == null) {
       return Optional.of("-html");
     } else {
-      return Optional.of("-html:" + stylesheet.asFile().getAbsolutePath());
+      return Optional.of("-html:" + getStylesheet().asFile().getAbsolutePath());
     }
   }
 
@@ -53,7 +55,34 @@ public class SpotBugsHtmlReport extends SpotBugsReport {
 
   @Override
   public TextResource getStylesheet() {
-    return stylesheet.getOrNull();
+    if (stylesheet.isPresent()) {
+      return stylesheet.get();
+    } else if (stylesheetPath.isPresent()) {
+      return resolve(stylesheetPath.get());
+    }
+
+    return null;
+  }
+
+  private TextResource resolve(String path) {
+    Optional<File> spotbugsJar =
+        getTask().getProject().getConfigurations().getByName("spotbugs")
+            .files(
+                dependency ->
+                    dependency.getGroup().equals("com.github.spotbugs")
+                        && dependency.getName().equals("spotbugs"))
+            .stream()
+            .findFirst();
+    if (spotbugsJar.isPresent()) {
+      return getTask()
+          .getProject()
+          .getResources()
+          .getText()
+          .fromArchiveEntry(spotbugsJar.get(), path);
+    } else {
+      throw new InvalidUserDataException(
+          "The dependency on SpotBugs not found in 'spotbugs' configuration");
+    }
   }
 
   @Override
@@ -63,23 +92,6 @@ public class SpotBugsHtmlReport extends SpotBugsReport {
 
   @Override
   public void setStylesheet(@Nullable String path) {
-    Optional<File> spotbugsJar =
-        getTask().getProject().getConfigurations().getByName("spotbugs")
-            // FIXME this operation probably evaluates the spotbugs configuration, that may make the
-            // build slow
-            .files(
-                dependency ->
-                    dependency.getGroup().equals("com.github.spotbugs")
-                        && dependency.getName().equals("spotbugs"))
-            .stream()
-            .findFirst();
-    if (spotbugsJar.isPresent()) {
-      TextResource textResource =
-          getTask().getProject().getResources().getText().fromArchiveEntry(spotbugsJar.get(), path);
-      setStylesheet(textResource);
-    } else {
-      throw new InvalidUserDataException(
-          "The dependency on SpotBugs not found in 'spotbugs' configuration");
-    }
+    stylesheetPath.set(path);
   }
 }

